Delay ActionDispatch::Response configuration to load-time

qnighy opened the pull request:
### Summary

I encountered an issue similar to #33283.

In a rails app which has just upgraded to Rails 6.0.0 from 5.2.3, I turned on `Rails.application.config.action_dispatch.return_only_media_type_on_content_type = false` config but it wasn't propagated to `ActionDispatch::Response.return_only_media_type_on_content_type`. I checked if [the initializer]((https://github.com/rails/rails/blob/v6.0.0/actionpack/lib/action_dispatch/railtie.rb#L41-L56)) is loaded too early from any third-party library (by inserting `raise "foo"` there and then running `bin/rails c`) but every row of the backtrace was in rails.

In ActionMailer [the configuration is delayed to load-time](https://github.com/rails/rails/blob/v6.0.0/actionmailer/lib/action_mailer/railtie.rb#L40-L54), so I guess ActionDispatch should also do that in [its initializer](https://github.com/rails/rails/blob/v6.0.0/actionpack/lib/action_dispatch/railtie.rb#L41-L56).

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

amatsuda commented:
Are you sure we have `:action_dispatch` load_hook?

qnighy commented:
> Are you sure we have `:action_dispatch` load_hook?

No, you were right. Thanks!

amatsuda commented:
Well, I don't think we're willing to add a new `on_load` hook until we're seeing strong reason for that.

So, what was the "third-party library" that loads AD before invoking :action_controller on_load?
Do you think it's difficult to fix your problem there in that library?

qnighy commented:
@amatsuda I mean I could find no such library.

qnighy commented:
I guess we could instead use `:before`/`:after` options of `Rails::Initializable::ClassMethods#initializer` to ensure `action_dispatch.configure` is run after `load_config_initializers`?

amatsuda commented:
> I mean I could find no such library

Oh, I'm sorry! I was not indeed reading what you wrote, obviously...

TBH I'm still wondering how you can configure an app not to propagate `Rails.application.config.action_dispatch.return_only_media_type_on_content_type` to `ActionDispatch::Response.return_only_media_type_on_content_type`.
Where in the app are you configuring `Rails.application.config.action_dispatch.return_only_media_type_on_content_type` value?

qnighy commented:
> Where in the app are you configuring `Rails.application.config.action_dispatch.return_only_media_type_on_content_type` value?

In `config/initializers/new_framework_defaults_6_0.rb`. I uncommented out the corresponding line in the file, then I observed the following situation:

```
$ bin/rails c
Loading development environment (Rails 6.0.0)
[1] pry(main)> Rails.application.config.action_dispatch.return_only_media_type_on_content_type
=> false
[2] pry(main)> ActionDispatch::Response.return_only_media_type_on_content_type
=> true
[3] pry(main)> exit
```

amatsuda commented:
That sounds weird.

I couldn't reproduce the situation on my side with a vanilla 5.2 => 6.0 app (both in development and production env), so there may be something in your app that triggers the AD initializer before reaching at the n_f_d_6_0 initializer file?

qnighy commented:
I tried to reproduce it in a vanilla rails app and it actually reproduced in my environment. Really weird...

amatsuda commented:
Ahh, OK, it's reproducing.

I confirm this is a bug. Let us see what's the best way to fix this...

qnighy commented:
Okay, in the meantime I'll configure the flag in `config/application.rb`.

y-yagi commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -39,20 +39,23 @@ class Railtie < Rails::Railtie # :nodoc:
-
-      ActionDispatch.test_app = app
+      config = self.config
+      ActiveSupport.on_load(:action_dispatch) do
```
Hmm. Is this approach really works expected? 
`ActionDispatch` is already loaded in this context. So `ActiveSupport.on_load(:action_dispatch)` executes immediately I think.

eugeneius replied:
Yeah, this would need to be a load hook for `ActionDispatch::Response` in order to be effective. Unfortunately, `ActionDispatch::Response` is eager loaded at boot time too:

https://github.com/rails/rails/blob/9680984cf8ce608294e8ee161d40eb87d7426a5a/actionpack/lib/action_controller.rb#L5
https://github.com/rails/rails/blob/9680984cf8ce608294e8ee161d40eb87d7426a5a/actionpack/lib/action_controller/metal/live.rb#L3

Maybe we can fix `ActionDispatch::Response` to autoload properly and add a hook.

eugeneius replied:
I opened https://github.com/rails/rails/pull/37228 to fix the autoload problem with `ActionDispatch::Response`.

eugeneius commented:
This test was meant to cover the case of an upgraded application:

https://github.com/rails/rails/blob/2b388c9648f992a81aff97e7d6f3cf8f3268c8ab/railties/test/application/configuration_test.rb#L2471-L2481

Unfortunately it has the same problem as https://github.com/rails/rails/pull/36946: it sets the old default value instead of the new one, so it passes even though the configuration has no effect.

Flipping the value reproduces the problem:

```patch
diff --git a/railties/test/application/configuration_test.rb b/railties/test/application/configuration_test.rb
index 0a6b95c6a0..6650c2c469 100644
--- a/railties/test/application/configuration_test.rb
+++ b/railties/test/application/configuration_test.rb
@@ -2472,12 +2472,12 @@ class ::DummySerializer < ActiveJob::Serializers::ObjectSerializer; end
       remove_from_config '.*config\.load_defaults.*\n'

       app_file "config/initializers/new_framework_defaults_6_0.rb", <<-RUBY
-        Rails.application.config.action_dispatch.return_only_media_type_on_content_type = true
+        Rails.application.config.action_dispatch.return_only_media_type_on_content_type = false
       RUBY

       app "development"

-      assert_equal true, ActionDispatch::Response.return_only_media_type_on_content_type
+      assert_equal false, ActionDispatch::Response.return_only_media_type_on_content_type
     end

     test "ActionMailbox.logger is Rails.logger by default" do
```

eugeneius commented:
@qnighy are you interested in finishing this? I think the remaining work is:

1. Rebase to pick up https://github.com/rails/rails/pull/37228.
2. Fix the existing test as shown in https://github.com/rails/rails/pull/37112#issuecomment-532432648, which will make it fail.
3. Move the load hook to `actionpack/lib/action_dispatch/http/response.rb`.
4. Wrap just the three lines that reference `ActionDispatch::Response` in the hook.

qnighy commented:
@eugeneius sorry for being absent and thanks for the detailed guide! I'll try to make it in several days.

qnighy committed:
Delay ActionDispatch::Response configuration to load-time

It fixes the problem in propagating return_only_media_type_on_content_type
and fixes the corresponding test being ineffective.

The mentioned test addes the following line:
...config.action_dispatch.return_only_media_type_on_content_type = true
to the config and checks if it takes effect. However, in this scenario,
the value is already true before this line.
Moreover, the users are supposed to flip this from true to false in real
situations.

This commit flips the config in the test, making it to fail as
expected. The next commit will fix the failure.

In order for return_only_media_type_on_content_type to appropriately
take effect on ActionDispatch::Response, we want to know when
ActionDispatch::Response is loaded.
As load hooks for ActionDispatch would be too broad, the appropriate
registry is for ActionDispatch::Response itself.

Looking into other examples, a hook name is a full class name in
snake case with `_base` suffix omitted, if any. Therefore, in this case,
:action_dispatch_response seems appropriate.

qnighy commented:
@eugeneius finished! Should I further squash the commits to cleanup histories, or keep the history as-is?

abhaynikam commented on railties/test/application/configuration_test.rb:
```diff
@@ -2472,12 +2472,12 @@ class ::DummySerializer < ActiveJob::Serializers::ObjectSerializer; end
       app "development"
 
-      assert_equal true, ActionDispatch::Response.return_only_media_type_on_content_type
+      assert_equal false, ActionDispatch::Response.return_only_media_type_on_content_type
```
I think you should use `assert_not` here instead of comparing booleans if that works for you.

```suggestion
      assert_not ActionDispatch::Response.return_only_media_type_on_content_type
```

qnighy replied:
I just followed the convention in the existing line, but I guess the literal `false` (formerly `true`) is intentional, indicating that we're checking if the configuration value is properly set.

eugeneius replied:
In this case I think it's more important to be consistent with the two preceding tests:

https://github.com/rails/rails/blob/9a62e95f4a61ec010a146ade9c705909ed82c8ec/railties/test/application/configuration_test.rb#L2459

https://github.com/rails/rails/blob/9a62e95f4a61ec010a146ade9c705909ed82c8ec/railties/test/application/configuration_test.rb#L2468

eugeneius commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -42,9 +42,11 @@ class Railtie < Rails::Railtie # :nodoc:
+      ActiveSupport.on_load(:action_dispatch_response) do
+        ActionDispatch::Response.default_charset = app.config.action_dispatch.default_charset || app.config.encoding
+        ActionDispatch::Response.default_headers = app.config.action_dispatch.default_headers
+        ActionDispatch::Response.return_only_media_type_on_content_type = app.config.action_dispatch.return_only_media_type_on_content_type
```
This block will run in the context of `ActionDispatch::Response`, and it's conventional to use `self` to refer to it:

https://guides.rubyonrails.org/v6.0.0/engines.html#example-3

```suggestion
        self.return_only_media_type_on_content_type = app.config.action_dispatch.return_only_media_type_on_content_type
```

qnighy replied:
Changed!

eugeneius commented:
Please do squash your commits. 👍

eugeneius commented:
Could you squash the two remaining commits into one? Otherwise this now looks good to me.

qnighy commented:
Sure!

ghiculescu commented:
> > Where in the app are you configuring `Rails.application.config.action_dispatch.return_only_media_type_on_content_type` value?
> 
> In `config/initializers/new_framework_defaults_6_0.rb`. I uncommented out the corresponding line in the file, then I observed the following situation:
> 
> ```
> $ bin/rails c
> Loading development environment (Rails 6.0.0)
> [1] pry(main)> Rails.application.config.action_dispatch.return_only_media_type_on_content_type
> => false
> [2] pry(main)> ActionDispatch::Response.return_only_media_type_on_content_type
> => true
> [3] pry(main)> exit
> ```

We are getting these results ^ when upgrading our app from Rails 5.2x to 6.0.3.2.

I created a draft PR here thinking that it was an oversight in a test: https://github.com/rails/rails/pull/39865, but the tests are passing.

Is anyone else seeing this issue when upgrading?

alxndrdude commented:
I can confirm this behaviour.

The `ActiveSupport.on_load(:action_dispatch_response)` hook loads before the `new_framework_defaults_6_0.rb`.
However, `config/application.rb` executes before `ActiveSupport.on_load(:action_dispatch_response)` hook.

That mean
- setting `Rails.application.config.action_dispatch.return_only_media_type_on_content_type` in `new_framework_defaults_6_0.rb` does not have any effect. (E.g. You would expect to get rid of the Deprecation warning, if `Rails.application.config.action_dispatch.return_only_media_type_on_content_type = false`, but this is not the case.)
- setting `config.action_dispatch.return_only_media_type_on_content_type` in `application.rb` does have an effect


Sorry for not providing a proper test example.
I only wanted to confirm the observation of @ghiculescu .


PS: upgraded from rails v5.2.4.4 to v6.0.3.4

laurmurclar commented:
I'm seeing the same behavior as @alxndrdude and @ghiculescu - should we create an issue? 
I've gotten around it by renaming the initializer to `1_new_framework_defaults_6_0.rb` so that it runs earlier than the action dispatch initializer, which uses the config value: 
https://github.com/rails/rails/blob/c04aff675d4cd3e8a54e0008dbfca10a8832a414/actionpack/lib/action_dispatch/railtie.rb#L48

rafaelfranca commented:
> I'm seeing the same behavior as @alxndrdude and @ghiculescu - should we create an issue? Or maybe re-open #33283?

The problem is in your app. It could be either a gem or a `ActionController::Base` reference that you have in an initializer or a file that is required at boot time.

laurmurclar commented:
Ah okay. It doesn't seem to be in our files, so it must be a gem. Thanks!
